### ISSUE TITLE
viafree.no and related domains

### ISSUE THREAD

**Issue by ghajini** [jspenguin2017/uBlockProtector#110]
did you just ask users on `viafree.no`
![screenshot 204](https://cloud.githubusercontent.com/assets/20338483/24330762/4432c79c-1243-11e7-95c3-76d8ef8cbfc2.png)

**Comment by uBlock-user** [comment]
Yes it's from the script. You can disable experimental solutions from here - https://x01x012013.github.io/AdBlockProtector/settings.html

**Comment by X01X012013** [comment]
Yes, when I was porting AAK Script, I noticed that the fix for `"viafree.no", "viafree.dk", "viafree.se", "tvplay.skaties.lv", "play.tv3.lt", "tv3play.tv3.ee"` doesn't seem to work. I'm not sure if I am Geo Locked, can you provide feedback on it? If the fix doesn't work at all, I might as well remove it.

**Comment by ghajini** [comment]
no ...instead i got antiadblock notice
![screenshot 206](https://cloud.githubusercontent.com/assets/20338483/24332410/bdd65e36-1263-11e7-9deb-478094a596c8.png)

**Comment by X01X012013** [comment]
Well, let's update our solution.

**Comment by X01X012013** [comment]
The old solution is now commented out, since it doesn't do anything useful and the `confirm` can be annoying.

**Comment by ghajini** [comment]
what can be done for this ?

**Comment by X01X012013** [comment]
@ghajini I'm not sure. Video sites are usually hard to deal with.

**Comment by uBlock-user** [comment]
Killing all setTimeout calls kills the anti adblock notice along witht he video player, so if you can find out which one to to kill precisely, then this will work.

**Comment by X01X012013** [comment]
@uBlock-user Thanks for the report, I'll have a look.

**Comment by uBlock-user** [comment]
You can log all setTimeout calls via  `##script:inject(setTimeout-logger.js)` and then get the list from the console.

**Comment by X01X012013** [comment]
@uBlock-user Unfortunately, there is only one function that is passed to `setTimeout`: 
```JavaScript
function d(){var d=[],f=arguments.length,g=!a||a&&a.deep!==!1;for(c&&j(c)&&c.apply(this,arguments);f--;)d[f]=g?e.wrap(a,arguments[f]):arguments[f];try{return b.apply(this,d)}catch(h){throw e.F(),e.captureException(h,a),h}}
```
Unminified: 
```JavaScript
function d() {
    var d = [],
        f = arguments.length,
        g = !a || a && a.deep !== !1;
    for (c && j(c) && c.apply(this, arguments); f--;) d[f] = g ? e.wrap(a, arguments[f]) : arguments[f];
    try {
        return b.apply(this, d)
    } catch (h) {
        throw e.F(), e.captureException(h, a), h
    }
}
```

**Comment by uBlock-user** [comment]
Well what can you do about it ?

**Comment by X01X012013** [comment]
@uBlock-user I'm not sure.

**Comment by uBlock-user** [comment]
I found more multiple setTimeout calls in the inline-script and yet only one is being sent ? That's certainly wrong.

**Comment by X01X012013** [comment]
@uBlock-user What did you get from the logger?

**Comment by uBlock-user** [comment]
same as yours, only one appears to be logged, while there are more than 10 in the inlinescript

**Comment by X01X012013** [comment]
As a side note, this is the old solution that doesn't seem to work anymore: 

<details>

```JavaScript
if (a.domCmp(["viafree.no", "viafree.dk", "viafree.se", "tvplay.skaties.lv", "play.tv3.lt", "tv3play.tv3.ee"])) {
    //(Experimental) Replace player on load
    const handler = function () {
        //Find player
        const elem = a.$("#video-player");
        if (elem.length === 0) {
            a.win.setTimeout(handler, 1000);
            return;
        }
        //Find ID
        let videoID = elem.attr("poster").split("/");
        videoID = videoID[videoID.length - 2];
        if (!videoID) {
            a.win.setTimeout(handler, 1000);
            return;
        }
        //Request data JSON
        GM_xmlhttpRequest({
            method: "GET",
            url: "http://playapi.mtgx.tv/v3/videos/stream/" + videoID,
            onload: function (result) {
                parser(result.responseText);
            }
        });
    };
    const parser = function (data) {
        //Parse response
        let streams;
        try {
            const parsedData = JSON.parse(data);
            streams = parsedData.streams
        } catch (err) {
            a.config.debugMode && a.out.err("AdBlock Protector failed to find video URL! ");
            //Activate hander again
            a.win.setTimeout(handler, 1000);
            return;
        }
        //Check source and type
        let sources = [], types = [];
        if (streams.high && streams.high !== "") {
            sources.push(streams.high);
            types.push("video/mp4");
        } else if (streams.hls && streams.hls !== "") {
            sources.push(streams.hls);
            types.push("application/x-mpegURL");
        } else if (streams.medium && streams.medium !== "") {
            sources.push(streams.medium);
            types.push(streams.medium.startsWith("rtmp") ? "rtmp/mp4" : "application/f4m+xml");
        } else {
            a.config.debugMode && a.out.err("AdBlock Protector failed to find video URL! ");
            return;
        }
        //Replace player
        a.videoJS.init();
        const height = a.$("#video-player").height();
        const width = a.$("#video-player").width();
        a.$("#video-player").after(a.videoJS(sources, types, width, height)).remove();
        //Activate hander again
        a.win.setTimeout(handler, 1000);
    };
    if (a.config.allowExperimental && a.win.confirm("AdBlock Protector says: \nThis fix is experimental " +
"and might not work, would you want to try it anyway? ")) {
        //Can't seem to find real source of the stream, Geo Lock in my area?
        handler();
    }
}
```

</details>

**Comment by uBlock-user** [comment]
the anti adblock notice is ran before the player intialises, that's why it won't work.

**Comment by X01X012013** [comment]
@szymon1118 We need some help fixing this video site, can you have a look please?

**Comment by szymon1118** [comment]
@X01X012013 I'm sorry but I was busy and I didn't have much time to look into it. I think that I'm Geolocked on this site because I translated the response (https://playapi.mtgx.tv/v3/videos/stream/828252) from one of the videos and it says that I can't access this website outside from Norway. I also noticed that this website is related to `viasatsport.fi` because when you have a look at variable `__CONFIG__` on `viasatsport.fi` you can extract url which probably they use to detect country/region: https://content.viaplay.se/regionlookup and a favicon is same or similar to `viafree.no` favicon also looks like they are geolocked their API so I suppose that people from allowed countries are still able to use current solution but correct me If I said something wrong.

**Comment by X01X012013** [comment]
@szymon1118 So that means our old solution DOES work, it just couldn't load media JSON because of GeoLock... 
```JavaScript
if (a.domCmp(["viafree.no", "viafree.dk", "viafree.se", "tvplay.skaties.lv", "play.tv3.lt", "tv3play.tv3.ee"])) {
    //(Experimental) Replace player on load
    const handler = function () {
        //Find player
        const elem = a.$("#video-player");
        if (elem.length === 0) {
            a.win.setTimeout(handler, 1000);
            return;
        }
        //Find ID
        let videoID = elem.attr("poster").split("/");
        videoID = videoID[videoID.length - 2];
        if (!videoID) {
            a.win.setTimeout(handler, 1000);
            return;
        }
        //Request data JSON
        GM_xmlhttpRequest({
            method: "GET",
            url: "http://playapi.mtgx.tv/v3/videos/stream/" + videoID,
            onload: function (result) {
                parser(result.responseText);
            }
        });
    };
    const parser = function (data) {
        //Parse response
        let streams;
        try {
            const parsedData = JSON.parse(data);
            streams = parsedData.streams
        } catch (err) {
            a.config.debugMode && a.out.err("AdBlock Protector failed to find video URL! ");
            //Activate hander again
            a.win.setTimeout(handler, 1000);
            return;
        }
        //Check source and type
        let sources = [], types = [];
        if (streams.high && streams.high !== "") {
            sources.push(streams.high);
            types.push("video/mp4");
        } else if (streams.hls && streams.hls !== "") {
            sources.push(streams.hls);
            types.push("application/x-mpegURL");
        } else if (streams.medium && streams.medium !== "") {
            sources.push(streams.medium);
            types.push(streams.medium.startsWith("rtmp") ? "rtmp/mp4" : "application/f4m+xml");
        } else {
            a.config.debugMode && a.out.err("AdBlock Protector failed to find video URL! ");
            return;
        }
        //Replace player
        a.videoJS.init();
        const height = a.$("#video-player").height();
        const width = a.$("#video-player").width();
        a.$("#video-player").after(a.videoJS(sources, types, width, height)).remove();
        //Activate hander again
        a.win.setTimeout(handler, 1000);
    };
    if (a.config.allowExperimental && a.win.confirm("AdBlock Protector says: \nThis fix is experimental " +
"and might not work, would you want to try it anyway? ")) {
        //Can't seem to find real source of the stream, Geo Lock in my area?
        handler();
    }
}
```

**Comment by lassekongo83** [comment]
I'm not GeoLocked on viafree.se. I get ad-reinsertion thanks to 2 EasyList whitelists.

`@@||fwmrm.net^*/admanager.js$domain=viafree.se
@@||fwmrm.net/ad/g/$script,domain=viafree.se`

If I block that with an $important filter I get the anti-adblock message instead of ad-reinsertion as expected. How can I test your script though? Also if there are any additional logs you need, let me know.

**Comment by szymon1118** [comment]
@lassekongo83 Can you test if this script works for you?
<details>

```javascript
if (a.domCmp(["viafree.no", "viafree.dk", "viafree.se", "tvplay.skaties.lv", "play.tv3.lt", "tv3play.tv3.ee"])) {
    //(Experimental) Replace player on load
    const handler = function () {
        //Find player
        const elem = a.$("#video-player");
        if (elem.length === 0) {
            a.win.setTimeout(handler, 1000);
            return;
        }
        //Find ID
        let videoID = a.win.vfAvodpConfig.videoId;
        if (!videoID) {
            a.win.setTimeout(handler, 1000);
            return;
        }
        //Request data JSON
        GM_xmlhttpRequest({
            method: "GET",
            url: "http://playapi.mtgx.tv/v3/videos/stream/" + videoID,
            onload: function (result) {
                parser(result.responseText);
            }
        });
    };
    const parser = function (data) {
        //Parse response
        let streams;
        try {
            const parsedData = JSON.parse(data);
            streams = parsedData.streams
        } catch (err) {
            a.config.debugMode && a.out.err("AdBlock Protector failed to find video URL! ");
            //Activate hander again
            a.win.setTimeout(handler, 1000);
            return;
        }
        //Check source and type
        let sources = [], types = [];
        if (streams.high && streams.high !== "") {
            sources.push(streams.high);
            types.push("video/mp4");
        } else if (streams.hls && streams.hls !== "") {
            sources.push(streams.hls);
            types.push("application/x-mpegURL");
        } else if (streams.medium && streams.medium !== "") {
            sources.push(streams.medium);
            types.push(streams.medium.startsWith("rtmp") ? "rtmp/mp4" : "application/f4m+xml");
        } else {
            a.config.debugMode && a.out.err("AdBlock Protector failed to find video URL! ");
            return;
        }
        //Replace player
        a.videoJS.init();
        const height = a.$("#video-player").height();
        const width = a.$("#video-player").width();
        a.$("#video-player").after(a.videoJS(sources, types, width, height)).remove();
        //Activate hander again
        a.win.setTimeout(handler, 1000);
    };
    if (a.config.allowExperimental && a.win.confirm("AdBlock Protector says: \nThis fix is experimental " +
"and might not work, would you want to try it anyway? ")) {
        //Can't seem to find real source of the stream, Geo Lock in my area?
        handler();
    }
}
```
</details>

**Comment by lassekongo83** [comment]
With that one I get an error message: "No compatible source was found for this video."

**Comment by szymon1118** [comment]
@X01X012013 I think we should include additional script in this case to handle m3u8 `videojs.hls.min.js`. Can you have a look?
Test link I got via proxy: 
<details>

You can test vjs video here: http://videojs.github.io/videojs-contrib-hls/

https://mtgxno04-vh.akamaihd.net//i//open//20xxxxxx//2036xxxx//203657xx//20365759//2e465479f94fe8d1e4dd92b062ce885c_20365759_072470_20170331T075553_,200,512,908,1400,2000,3500,.mp4.csmil//master.m3u8?__b__=300&hdnts=st=1491247785~exp=3638731432~acl=\/*~hmac=89a79f634009a64e10139733648e74f38a5818bd6ae0718bf28302fe0596b5c7&cc1=name=Norsk~default=yes~forced=no~lang=no~uri=https://substitch.play.mtgx.tv/subtitle/convert/vtt?source=https://cdn-subtitles-mtgx-tv.akamaized.net/pitcher/20xxxxxx/2036xxxx/203657xx/20365759/6430163904/unified.subs.20365759_no.vtt&output=m3u8&cc2=name=Norsk%20for%20h%C3%83%C2%B8rselshemmede~default=no~forced=no~lang=no~uri=https://substitch.play.mtgx.tv/subtitle/convert/vtt?source=https://cdn-subtitles-mtgx-tv.akamaized.net/pitcher/20xxxxxx/2036xxxx/203657xx/20365759/6430163904/unified.subs.20365759_no_sdh.vtt&output=m3u8
</details>

**Comment by X01X012013** [comment]
@lassekongo83 
Can you dump the server response from `onload`? 
```JavaScript
        GM_xmlhttpRequest({
            method: "GET",
            url: "http://playapi.mtgx.tv/v3/videos/stream/" + videoID,
            onload: function (result) {
                parser(result.responseText);
            }
        });
```

Which country are you in? Maybe I can find a VPN.

**Comment by X01X012013** [comment]
@szymon1118 Ah... I see... Forget about what I said before. I think our `a.videoJS` can't handle those format... Do you have a link to the CDN that distributes those libraries?

**Comment by szymon1118** [comment]
@X01X012013 cdnjs.cloudflare.com/ajax/libs/videojs-contrib-hls/5.4.0/videojs-contrib-hls.min.js

**Comment by lassekongo83** [comment]
@X01X012013 Can you instruct me on how to that? :)

**Comment by ghajini** [comment]
@X01X012013 you can use this method described 
[url=https://www.reddit.com/r/norsk/comments/2dshka/tutorial_how_to_watch_nrktv_while_being_outisde/]method[/url]
i find it,just take proxy for norway from google

**Comment by X01X012013** [comment]
Sweet, I have Sweden in my list 😄 
I'll have a look. 
![image](https://cloud.githubusercontent.com/assets/7283682/24711809/42671ae2-19de-11e7-8579-ab346e5bd5cd.png)

**Comment by szymon1118** [comment]
@X01X012013 I also found a good proxy so you can simply use it like below :grinning: 
<details>

```javascript
//Request data JSON
GM_xmlhttpRequest({
	method: "GET",
	url: "http://www.sagkjeder.no/p/browse.php?u=http://playapi.mtgx.tv/v3/videos/stream/" + videoID,
	onload: function (result) {
		parser(result.responseText);
	}
});
```
</details>

**Comment by X01X012013** [comment]
Hi guys, I just committed version 6.143. The development solution is added to Script, it will only run if you are on Debug Mode. Debug Mode can be turned on in the settings page, remember to turn it off when you are done, it slows down Script while making it less reliable, it's really only for development purposes. 
I tried loading the HLS parser, it doesn't seem to work... I'm a bit busy today, I'll look into this later.

**Comment by szymon1118** [comment]
@X01X012013 I found a useful comment in this issue: https://github.com/videojs/videojs-contrib-hls/issues/1018#issuecomment-292098580 so you should add `withCredentials` like above :smile: and it should work.
<details>

```javascript
let html = `<video id="AdBlock_Protector_Video_Player" class="video-js vjs-default-skin" controls preload="auto" width="${width}" height="${height}" data-setup='{"withCredentials": true}'>`;
```
</details>

**Comment by X01X012013** [comment]
@szymon1118 Does that works for you or you haven't tested that yet? I'm still getting the same error...
![image](https://cloud.githubusercontent.com/assets/7283682/24765558/7fde7c26-1ab5-11e7-99dc-6e0fa8735df6.png)

**Comment by szymon1118** [comment]
@X01X012013 Sorry but I was able to play a video but a videoJS player had a bug because a play button was in the top-left corner and later when I tried to play once again I got "No compatible source..." error so I need to investigate what it is going on.

**Comment by X01X012013** [comment]
Hi, I added some logging to the debug code, the response will now be dumped to the console. 

Test URL: https://www.viafree.no/programmer/underholdning/paradise-hotel/sesong-9/episode-16
Response is a HTML, not JSON, could not parse. 

Test URL: https://www.viafree.no/programmer/underholdning/paradise-hotel/sesong-9/episode-14
Response is JSON, `streams.hls` is present, file type seems to be `m3u8`, could not play the file. 

Test URL: https://www.viafree.no/programmer/underholdning/paradise-hotel/sesong-9/episode-13
Response is JSON, `streams.hls` is present, file type seems to be `m3u8`, player initialized, but stuck at loading. The style of the player is a bit off, probably conflict with original styles. Note that by default, the play button of VideoJS is on the left-up corner. 

Known issues: 
When the response could not be parsed, the script will abort and will not trigger if another video is opened. The trigger of the script is currently the presence of the original player, if we failed to replace it, we could not schedule the next check, we might need to use something else as the trigger... Or add a flag to the original player?
